### ISSUE TITLE
"Any object created by object A is object A's responsibility to destroy." + API

### ISSUE THREAD

**Issue by dcporter** [sproutcore/sproutcore#869]
There's an underabundance of attention paid in most SproutCore projects (NTM the framework itself) to the .destroy end of the object lifecycle. Object creation is well-understood, but destruction is generally assumed to take care of itself. GC in browsers is such that this is often true (many SC objects can be GC'd without being explicitly destroyed), but not often enough to prevent leakage.

The golden rule of object lifecycle management is that "Any object created by object A is object A's responsibility to destroy". There's no support whatsoever in SC for this, however, so any class that chooses to follow it has to implement its own tracking and destroying.

I'd like to implement a simple API in SC.Object (or mixed in there) for tracking object ownership and triggering destruction. The goal is to make it as simple as possible to obey the golden rule by providing an "ownership" concept whereby an object can declare that it created another object and will take responsibility for destroying it. Owned objects must also be aware of this, so that if they're destroyed at some other time, they're not improperly retained by the owner.

Here's an off-the-top-of-my-head API design:
- property **ownerObject**: the object's owner. This should be set via the .create({}) hash if possible, or immediately afterwards via takeOwnershipOfObject.
- method **takeOwnershipOfObject**(obj): receiver object takes ownership of obj. If obj is already owned by somebody else, an exception is thrown. (Changing owners would be bad.) This would add the object to an internal list of owned objects. You'd call it from the receiver when it created the object. If the object is going to live on the receiver as a key (e.g. ScrollView#horizontalScrollerView) then you can optionally pass in a key. When the owner object is destroyed, it destroys all owned objects.
- property **doesOwnObject**(obj): returns boolean indicating whether the receiver is the owner of obj.
- method **destroyOwnedObject**(obj): destroys obj, as long as it's owned by the receiver. Otherwise does nothing. You can call this any time you want to destroy an object you created (e.g. if you're swapping views in a ContainerView).
- method **ownedObjectWasDestroyed**(obj): when an object is destroyed, it needs to alert its owner that it no longer needs tracking. This removes the owned object from internal ownership tracking.

Finally, #destroy (or #destroyMixin) would be updated to destroy all remaining owned objects, and to alert its owner that it's done for.

The three use case examples that I can think of off the top of my head include:
- Every childView passed in as a class would be automatically destroyed, while childViews that arrived instantiated would be left alone. (Currently all children are destroyed, regardless.)
- SC.ContainerView lets you pass in classes, which it will instantiate for you. At that point, it would call this.takeOwnershipOfObject. When swapping that view out for a new one, the container would just call this.destroyOwnedObject, which only destroys objects that it's supposed to. Finally, when the container is destroyed, no further effort would be required to ensure that its final nowShowing is also destroyed.
- SC.ScrollView currently creates, but does not destroy, instances of ScrollerView and ContainerView. With SC.Ownable, it would take ownership of those views at create time, causing them to be automatically destroyed at destroy time. In rare circumstances, ScrollView doesn't create these views itself; in these cases, the views would be properly left intact.

(Objects would still be responsible for nulling out their own properties on destroy, e.g. ScrollView's container and scroller views. It would be easy to create a declarative API for this as well, but I'm on the fence about the need. Also, I think that once objects are destroying themselves properly, nulling out properties will be less important, as isolated groups of destroyed objects should be GC'd together.)

A) How ... readable is that-all?
B) Thoughts?
C) Is this better than just shouting about the golden rule and doing our best to follow it?

Thanks,
Dave

**TL;DR**: The golden rule of object lifecycles says "If object A creates object B, object A is in charge of making sure .destroy() is called on object B at some point". Bulleted above is an off-the-cuff API proposal introducing & supporting the concept of object ownership to make it easier to follow this rule.
